File: scripts/collect_rir_delegated_monthly.go
package main

import (
	"context"
	"crypto/sha256"
	"encoding/csv"
	"encoding/hex"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	defaultBaseURL = "https://ftp.ripe.net/pub/stats/ripencc/nro-stats"
	userAgent      = "asn-mismatch-pipeline/0.1 (+go monthly rir delegated snapshots)"
)

var sourceCandidates = []string{"nro-delegated-stats", "combined-stat"}

type job struct {
	Month        string
	SnapshotDate string
}

type result struct {
	Month        string
	SnapshotDate string
	SourceName   string
	SourceURL    string
	Path         string
	SHA256       string
	Bytes        int64
	LineCount    int64
	Status       string
	FetchedAt    string
	Err          error
}

func main() {
	var (
		months     = flag.Int("months", 60, "number of month-end snapshots to download")
		endMonth   = flag.String("end-month", lastCompletedMonth(time.Now().UTC()), "inclusive end month in YYYY-MM")
		workers    = flag.Int("workers", 4, "number of concurrent downloads")
		outputDir  = flag.String("output-dir", "data/raw/registry/delegated_monthly_go", "output directory")
		timeoutSec = flag.Int("timeout-sec", 180, "per-request timeout in seconds")
		baseURL    = flag.String("base-url", defaultBaseURL, "NRO stats base URL")
	)
	flag.Parse()

	if *months <= 0 {
		fatalf("--months must be positive")
	}
	if *workers <= 0 {
		fatalf("--workers must be positive")
	}
	if err := os.MkdirAll(*outputDir, 0o755); err != nil {
		fatalf("create output dir: %v", err)
	}

	client := &http.Client{Timeout: time.Duration(*timeoutSec) * time.Second}
	jobs := make(chan job)
	results := make(chan result)

	var wg sync.WaitGroup
	for i := 0; i < *workers; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			for item := range jobs {
				results <- downloadMonth(client, *baseURL, *outputDir, item)
			}
		}(i)
	}

	go func() {
		for _, month := range monthSequence(*endMonth, *months) {
			jobs <- job{Month: month, SnapshotDate: monthEndDate(month)}
		}
		close(jobs)
		wg.Wait()
		close(results)
	}()

	var rows []result
	var failed int
	for row := range results {
		rows = append(rows, row)
		if row.Err != nil {
			failed++
			fmt.Fprintf(os.Stderr, "[ERROR] %s %s: %v\n", row.Month, row.SnapshotDate, row.Err)
			continue
		}
		fmt.Printf("[%s] %s -> %s (%d bytes, %d lines)\n", row.Status, row.Month, row.Path, row.Bytes, row.LineCount)
	}

	sort.Slice(rows, func(i, j int) bool {
		return rows[i].Month < rows[j].Month
	})
	if err := writeIndex(filepath.Join(*outputDir, "index.csv"), rows); err != nil {
		fatalf("write index: %v", err)
	}
	fmt.Printf("saved %d monthly delegated snapshots; failed=%d; index=%s\n", len(rows)-failed, failed, filepath.Join(*outputDir, "index.csv"))
	if failed > 0 {
		os.Exit(1)
	}
}

func downloadMonth(client *http.Client, baseURL, outputDir string, item job) result {
	finalPath := filepath.Join(outputDir, fmt.Sprintf("nro_delegated_stats_%s_%s.txt", item.Month, item.SnapshotDate))
	now := time.Now().UTC().Format(time.RFC3339)

	if info, err := os.Stat(finalPath); err == nil {
		sha, err := sha256File(finalPath)
		if err != nil {
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, Err: err, FetchedAt: now}
		}
		lines, err := countLines(finalPath)
		if err != nil {
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, Err: err, FetchedAt: now}
		}
		return result{
			Month: item.Month, SnapshotDate: item.SnapshotDate, Path: finalPath, SHA256: sha,
			Bytes: info.Size(), LineCount: lines, Status: "skipped_existing", FetchedAt: now,
		}
	}

	var lastErr error
	for _, sourceName := range sourceCandidates {
		url := strings.TrimRight(baseURL, "/") + "/" + item.SnapshotDate + "/" + sourceName
		tmpPath := finalPath + ".tmp"
		if err := fetchURL(client, url, tmpPath); err != nil {
			_ = os.Remove(tmpPath)
			lastErr = fmt.Errorf("%s: %w", sourceName, err)
			continue
		}
		lines, err := countLines(tmpPath)
		if err != nil {
			_ = os.Remove(tmpPath)
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, SourceName: sourceName, SourceURL: url, Err: err, FetchedAt: now}
		}
		if err := os.Rename(tmpPath, finalPath); err != nil {
			_ = os.Remove(tmpPath)
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, SourceName: sourceName, SourceURL: url, Err: err, FetchedAt: now}
		}
		sha, err := sha256File(finalPath)
		if err != nil {
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, SourceName: sourceName, SourceURL: url, Err: err, FetchedAt: now}
		}
		info, err := os.Stat(finalPath)
		if err != nil {
			return result{Month: item.Month, SnapshotDate: item.SnapshotDate, SourceName: sourceName, SourceURL: url, Err: err, FetchedAt: now}
		}
		return result{
			Month: item.Month, SnapshotDate: item.SnapshotDate, SourceName: sourceName, SourceURL: url,
			Path: finalPath, SHA256: sha, Bytes: info.Size(), LineCount: lines, Status: "ok", FetchedAt: now,
		}
	}

	return result{Month: item.Month, SnapshotDate: item.SnapshotDate, Err: lastErr, FetchedAt: now}
}

func fetchURL(client *http.Client, url, path string) error {
	ctx, cancel := context.WithTimeout(context.Background(), client.Timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", userAgent)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	out, err := os.Create(path)
	if err != nil {
		return err
	}
	defer out.Close()

	if _, err := io.CopyBuffer(out, resp.Body, make([]byte, 1024*1024)); err != nil {
		return err
	}
	return out.Sync()
}

func writeIndex(path string, rows []result) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	w := csv.NewWriter(f)
	defer w.Flush()
	header := []string{"month", "snapshot_date", "source_name", "source_url", "raw_evidence_path", "raw_evidence_sha256", "bytes", "line_count", "status", "fetched_at", "error"}
	if err := w.Write(header); err != nil {
		return err
	}
	for _, row := range rows {
		errText := ""
		if row.Err != nil {
			errText = row.Err.Error()
		}
		if err := w.Write([]string{
			row.Month,
			row.SnapshotDate,
			row.SourceName,
			row.SourceURL,
			row.Path,
			row.SHA256,
			fmt.Sprintf("%d", row.Bytes),
			fmt.Sprintf("%d", row.LineCount),
			row.Status,
			row.FetchedAt,
			errText,
		}); err != nil {
			return err
		}
	}
	return w.Error()
}

func sha256File(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.CopyBuffer(h, f, make([]byte, 1024*1024)); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

func countLines(path string) (int64, error) {
	f, err := os.Open(path)
	if err != nil {
		return 0, err
	}
	defer f.Close()

	var count int64
	buf := make([]byte, 1024*1024)
	for {
		n, err := f.Read(buf)
		for _, b := range buf[:n] {
			if b == '\n' {
				count++
			}
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return 0, err
		}
	}
	return count, nil
}

func lastCompletedMonth(now time.Time) string {
	year, month, _ := now.Date()
	month--
	if month == 0 {
		year--
		month = 12
	}
	return fmt.Sprintf("%04d-%02d", year, month)
}

func monthSequence(endMonth string, count int) []string {
	parts := strings.Split(endMonth, "-")
	if len(parts) != 2 {
		fatalf("invalid --end-month %q", endMonth)
	}
	var year, month int
	if _, err := fmt.Sscanf(endMonth, "%04d-%02d", &year, &month); err != nil || month < 1 || month > 12 {
		fatalf("invalid --end-month %q", endMonth)
	}

	months := make([]string, 0, count)
	for i := 0; i < count; i++ {
		months = append(months, fmt.Sprintf("%04d-%02d", year, month))
		month--
		if month == 0 {
			year--
			month = 12
		}
	}
	for i, j := 0, len(months)-1; i < j; i, j = i+1, j-1 {
		months[i], months[j] = months[j], months[i]
	}
	return months
}

func monthEndDate(month string) string {
	var year, monthNum int
	if _, err := fmt.Sscanf(month, "%04d-%02d", &year, &monthNum); err != nil {
		fatalf("invalid month %q", month)
	}
	lastDay := 31
	switch monthNum {
	case 4, 6, 9, 11:
		lastDay = 30
	case 2:
		lastDay = 28
		if year%400 == 0 || (year%4 == 0 && year%100 != 0) {
			lastDay = 29
		}
	}
	return fmt.Sprintf("%04d%02d%02d", year, monthNum, lastDay)
}

func fatalf(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

File: scripts/collect_rir_delegated_monthly.py
#!/usr/bin/env python3
"""Download monthly NRO/RIR delegated stats snapshots.

The RIPE NCC mirror publishes NRO-combined delegated statistics under
``pub/stats/ripencc/nro-stats/YYYYMMDD``.  One month-end snapshot is enough for
the raw registry baseline use case here because the file already contains the
five RIR views in a single exchange-format file.
"""

from __future__ import annotations

import argparse
import calendar
import csv
import gzip
import shutil
import sys
import tempfile
from datetime import date
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from pipeline_utils import ensure_dirs, load_config, relative_to_root, sha256_file, utc_now, write_json


BASE_URL = "https://ftp.ripe.net/pub/stats/ripencc/nro-stats"
USER_AGENT = "asn-mismatch-pipeline/0.1 (+monthly rir delegated snapshots)"
SOURCE_CANDIDATES = ("nro-delegated-stats", "combined-stat")


def last_completed_month(today: date) -> str:
    year = today.year
    month = today.month - 1
    if month == 0:
        year -= 1
        month = 12
    return f"{year:04d}-{month:02d}"


def month_sequence(end_month: str, count: int) -> list[str]:
    year, month = [int(part) for part in end_month.split("-")]
    months: list[str] = []
    for _ in range(count):
        months.append(f"{year:04d}-{month:02d}")
        month -= 1
        if month == 0:
            year -= 1
            month = 12
    return list(reversed(months))


def month_end_yyyymmdd(month: str) -> str:
    year, month_num = [int(part) for part in month.split("-")]
    day = calendar.monthrange(year, month_num)[1]
    return f"{year:04d}{month_num:02d}{day:02d}"


def source_url(snapshot_date: str, source_name: str) -> str:
    return f"{BASE_URL}/{snapshot_date}/{source_name}"


def fetch_to_path(url: str, path: Path, timeout: int) -> int:
    request = Request(url, headers={"User-Agent": USER_AGENT})
    with urlopen(request, timeout=timeout) as response:
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("wb") as f:
            shutil.copyfileobj(response, f, length=1024 * 1024)
        return int(response.headers.get("Content-Length") or path.stat().st_size)


def try_download(snapshot_date: str, tmp_path: Path, timeout: int) -> tuple[str, str, int]:
    errors: list[str] = []
    for source_name in SOURCE_CANDIDATES:
        url = source_url(snapshot_date, source_name)
        try:
            size = fetch_to_path(url, tmp_path, timeout)
            return source_name, url, size
        except HTTPError as exc:
            if tmp_path.exists():
                tmp_path.unlink()
            errors.append(f"{source_name}: HTTP {exc.code}")
        except (URLError, TimeoutError, OSError) as exc:
            if tmp_path.exists():
                tmp_path.unlink()
            errors.append(f"{source_name}: {exc}")
    raise RuntimeError("; ".join(errors))


def gzip_file(src: Path, dst: Path) -> None:
    with src.open("rb") as f_in, gzip.open(dst, "wb", compresslevel=6) as f_out:
        shutil.copyfileobj(f_in, f_out, length=1024 * 1024)


def count_data_lines(path: Path) -> int:
    count = 0
    with path.open("rb") as f:
        for line in f:
            if line.strip():
                count += 1
    return count


def write_index(path: Path, rows: list[dict[str, str | int]]) -> None:
    fieldnames = [
        "month",
        "snapshot_date",
        "source_name",
        "source_url",
        "raw_evidence_path",
        "raw_evidence_sha256",
        "bytes",
        "line_count",
        "status",
        "fetched_at",
    ]
    with path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def main() -> int:
    parser = argparse.ArgumentParser(description="Download monthly NRO/RIR delegated stats snapshots.")
    parser.add_argument("--config", type=Path, default=Path("configs/pipeline.yaml"))
    parser.add_argument("--months", type=int, default=60)
    parser.add_argument("--end-month", default=last_completed_month(date.today()), help="Inclusive YYYY-MM month.")
    parser.add_argument("--output-dir", type=Path, default=None)
    parser.add_argument("--timeout-sec", type=int, default=120)
    parser.add_argument("--compress", action="store_true", help="Store monthly files as local .gz files.")
    args = parser.parse_args()

    if args.months <= 0:
        raise ValueError("--months must be positive")

    config = load_config(args.config)
    output_dir = args.output_dir or Path(config["paths"]["raw_root"]) / "registry" / "delegated_monthly"
    ensure_dirs([output_dir])
    output_dir = output_dir.resolve()

    rows: list[dict[str, str | int]] = []
    errors: list[str] = []
    for month in month_sequence(args.end_month, args.months):
        snapshot_date = month_end_yyyymmdd(month)
        suffix = ".txt.gz" if args.compress else ".txt"
        final_path = output_dir / f"nro_delegated_stats_{month}_{snapshot_date}{suffix}"
        if final_path.exists():
            status = "skipped_existing"
            source_name = "unknown_existing"
            url = ""
            line_count = 0
            bytes_written = final_path.stat().st_size
        else:
            with tempfile.TemporaryDirectory(dir=output_dir) as tmp_dir:
                tmp_path = Path(tmp_dir) / f"{snapshot_date}.download"
                try:
                    source_name, url, _source_size = try_download(snapshot_date, tmp_path, args.timeout_sec)
                    line_count = count_data_lines(tmp_path)
                    if args.compress:
                        gzip_file(tmp_path, final_path)
                    else:
                        tmp_path.replace(final_path)
                    bytes_written = final_path.stat().st_size
                    status = "ok"
                except Exception as exc:
                    errors.append(f"{month} {snapshot_date}: {exc}")
                    print(f"[ERROR] {month} {snapshot_date}: {exc}", file=sys.stderr)
                    continue

        rows.append(
            {
                "month": month,
                "snapshot_date": snapshot_date,
                "source_name": source_name,
                "source_url": url,
                "raw_evidence_path": relative_to_root(final_path),
                "raw_evidence_sha256": sha256_file(final_path),
                "bytes": bytes_written,
                "line_count": line_count,
                "status": status,
                "fetched_at": utc_now(),
            }
        )
        print(f"[{status}] {month} -> {relative_to_root(final_path)}")

    index_path = output_dir / "index.csv"
    write_index(index_path, rows)
    log_path = output_dir / f"download_log_{utc_now().replace(':', '').replace('-', '')}.json"
    write_json(
        log_path,
        {
            "fetched_at": utc_now(),
            "source_base_url": BASE_URL,
            "months_requested": args.months,
            "months_saved": len(rows),
            "end_month": args.end_month,
            "output_dir": relative_to_root(output_dir),
            "index_path": relative_to_root(index_path),
            "errors": errors,
        },
    )
    print(f"saved {len(rows)} monthly delegated snapshots; index={relative_to_root(index_path)}")
    return 1 if errors else 0


if __name__ == "__main__":
    raise SystemExit(main())
